compiler-core/src/javascript/typescript.rs

```diff
@@ -0,0 +1,530 @@
+    /// low level types the rest of the type system are built up from. If there
+    /// is no built-in Typescript equivalent, the type is prefixed with "$gleam."
+    /// and the Gleam prelude namespace will be imported during the code emission.
+    ///
+    fn print_prelude_type(&self, name: &str, args: &[Arc<Type>]) -> Document<'static> {
+        match name {
+            "Nil" => "null".to_doc(),
+            "Int" | "Float" => "number".to_doc(),
+            "UtfCodepoint" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                "$gleam.UtfCodepoint".to_doc()
+            }
+            "String" => "string".to_doc(),
+            "Bool" => "boolean".to_doc(),
+            "BitString" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                "$gleam.BitString".to_doc()
+            }
+            "List" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                let arg0 = self.print(args.get(0).expect("print_prelude_type list"));
+                docvec!["$gleam.List<", arg0, ">"]
+            }
+            "Result" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                let arg_ok = self.print(args.get(0).expect("print_prelude_type result ok"));
+                let arg_err = self.print(args.get(1).expect("print_prelude_type result err"));
+                docvec!["$gleam.Result<", arg_ok, ", ", arg_err, ">"]
+            }
+            // Getting here sholud mean we either forgot a built-in type or there is a
+            // compiler error
+            name => panic!("{} is not a built-in type.", name),
+        }
+    }
+
+    /// Prints a "named" programmer-defined Gleam type into the Typescript
+    /// equivalent.
+    ///
+    fn print_type_app(
+        &self,
+        name: &str,
+        args: &[Arc<Type>],
+        module: &[String],
+    ) -> Document<'static> {
+        let name = format!("{}$", ts_safe_type_name(name.to_camel_case()));
+        let name = match module == self.current_module {
+            true => Document::String(name),
+            false => {
+                // If type comes from a seperate module, use that module's nam
+                // as a Typescript namespace prefix
+                let module_name = module
+                    .last()
+                    .expect("Typescript generator could not identify imported module name.");
+                let module_name = format!("${}", module_name);
+                docvec![Document::String(module_name), ".", Document::String(name),]
+            }
+        };
+        if args.is_empty() {
+            return name;
+        }
+        // If the App type takes arguments, pass them in as Typescript generics
+        let args = concat(Itertools::intersperse(
+            args.iter().map(|a| self.print(a)),
+            ", ".to_doc(),
+        ));
+        docvec![name, "<", args, ">"]
```

Comment:
It would be nice to construct this document so that it wraps nicely if it doesn't fit on one line, similar to how a function call works for JS generation.

Once we have it we could also reuse this nicely indenting document construction for printing Result and List from the stdlib.

Reply:
Will look at what Javascript does and move that here as well!